### PR TITLE
Implementing Rate Limiting with Resilience4j

### PR DESCRIPTION
**Pull request by saajn**
Please re-read your article a day or so after you finished writing it and work through this checklist to polish the quality:

- [x] the article explains *why* and not only *how*
- [x] the text is conversational ("we" instead of "you", easy language)
- [x] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [x] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [x] the text uses active voice instead of passive voice
- [x] names are spelled consistently throughout the article (uppercase / lowercase)
- [x] the article has a short intro before the first headline
- [x] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [x] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [x] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [x] code examples are introduced with a ":" and explained below the code example
- [x] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)
- [x] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

**Feel free to raise the PR even if the checklist hasn't been worked through, yet**, to take advantage of the preview that is generated for each pull request (see "deploy/netlify" under "Show all checks"). But please only flag it to be reviewed after going through the checklist above.

### CODE REVIEW

File: _posts/2020-07-19-rate-limiting-with-resilience4j.md [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+```
+RateLimiterEvent{type=SUCCESSFUL_ACQUIRE, rateLimiterName='flightSearchService', creationTime=2020-07-21T19:14:33.127+05:30}
+... other lines omitted ...
+RateLimiterEvent{type=FAILED_ACQUIRE, rateLimiterName='flightSearchService', creationTime=2020-07-21T19:14:33.186+05:30}
+```
+
+## RateLimiter Metrics
+
+Suppose after implementing client-side throttling we find that the response times of our APIs have increased. This is possible because our request processing threads sometimes wait to get permission before calling the remote operation. How can we find out how often we are throttling our requests? How can we ensure that the values we've set on the `RateLimiterConfig` are working well?
```
Comment (thombergs) [inline _posts/2020-07-19-rate-limiting-with-resilience4j.md:258]: I don't understand the motivation to check how often we're throttling our requests when the response times have increased. Can you rephrase? I'm especially struggling with the second sentence

File: _posts/2020-07-19-rate-limiting-with-resilience4j.md [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+```
+
+After running the rate-limited operation a few times, we display the captured metrics:
+
+```java
+Consumer<Meter> meterConsumer = meter -> {
+  String desc = meter.getId().getDescription();
+  String metricName = meter.getId().getName();
+  Double metricValue = StreamSupport.stream(meter.measure().spliterator(), false).
+    filter(m -> m.getStatistic().name().equals("VALUE")).
```
Comment (thombergs) [inline _posts/2020-07-19-rate-limiting-with-resilience4j.md]: I've never seen anyone format their code with the "." at the end of the line (even the code formatter in Jekyll had problems with that :). I've re-formatted it to put the "." at the start of the next line.

File: _posts/2020-07-19-rate-limiting-with-resilience4j.md [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+```
+
+The negative value for `resilience4j.ratelimiter.available.permissions` shows the number of permissions that have been reserved for requesting threads. In a real application, we would export the data to a monitoring system periodically and analyze it on a dashboard.
+
+## Gotchas and Good Practices When Implementing Client-side Rate Limiting
+
+**All calls to a given remote service should go through the same `RateLimiter` instance. For a given remote service the `RateLimiter` must be a singleton**. If we don't enforce this, some areas of our codebase may make a direct call to the remote service, bypassing the `RateLimiter`. To prevent this, the actual call to the remote service should be in a core, internal layer and other areas should use a rate-limited decorator exposed by the internal layer.
+
+How can we ensure that a new developer understands this intent in the future? Check out Tom's [article](https://reflectoring.io/java-components-clean-boundaries/) which shows one way of solving such problems by organizing the package structure to make such intents clear. Additionally, it shows how to enforce this by codifying the intent in ArchUnit tests.
```
Comment (thombergs) [inline _posts/2020-07-19-rate-limiting-with-resilience4j.md]: Nice, even I wouldn't have made that connection :)

File: _posts/2020-07-19-rate-limiting-with-resilience4j.md (hunk 1 above)
Comment (saajn) [reply _posts/2020-07-19-rate-limiting-with-resilience4j.md:258]: Haha, it was a blind spot I guess. I had never noticed that. I had read somewhere that when you have long statements it's better to break the line after the operator so that a developer reading knowing that it continues, something like

int sum = a + 
                 b +
                 ...

I guess I had carried that over to "." as well. And you are right - I noticed after you mentioned that nearly everyone has the "." on the next line. I'll do the same next time. :)

File: _posts/2020-07-19-rate-limiting-with-resilience4j.md (hunk 2 above)
Comment (saajn) [reply _posts/2020-07-19-rate-limiting-with-resilience4j.md]: Haha, it was a blind spot I guess. I had never noticed that. I had read somewhere that when you have long statements it's better to break the line after the operator so that a developer reading knowing that it continues, something like

int sum = a +
b +
...

I guess I had carried that over to "." as well. And you are right - I noticed after you mentioned that nearly everyone has the "." on the next line. I'll do the same next time. :)

File: _posts/2020-07-19-rate-limiting-with-resilience4j.md (hunk 1 above)
Comment (saajn) [reply _posts/2020-07-19-rate-limiting-with-resilience4j.md]: Agreed, couple of lines there were a bit confusing. I have made some changes to this and the next paragraph. Can you check if it reads better now?

Comment (thombergs) [reply _posts/2020-07-19-rate-limiting-with-resilience4j.md]: Yes, much clearer now! :)

File: _posts/2020-07-19-rate-limiting-with-resilience4j.md (hunk 2 above)
Comment (thombergs) [reply _posts/2020-07-19-rate-limiting-with-resilience4j.md]: Right, I put operators like + and = on the same line, too.
